File: c++/468validateIpAddress.cc
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <climits>
#include <stack>
#include <algorithm>
#include <cmath>
#include <set>
#include <unordered_map>
#include <list>
#include <unordered_set>
#include <map>
#include <set>
#include <functional>
#include <bitset>
#include <numeric>
#include <sstream>

using namespace std;

static int x = [] () {ios::sync_with_stdio(false); cin.tie(0); return 0;} ();
class Solution {
 public:
  string validIPAddress(string queryIP) {
    if (queryIP.size() == 0 || queryIP.back() == '.' || queryIP.back() == ':') return "Neither";
    istringstream ip(queryIP);
    string cur = "";
    int count = 0;

    auto ipv4 = [&] () {
      if (queryIP.find('.') == string::npos) return false;
      while (getline(ip, cur, '.')) {
        ++count;
        if (count > 4 || cur.empty() || cur.size() > 3 || (cur.size() > 1 && cur[0] == '0'))
          return false;
        int val = 0;
        for (const char c : cur) {
          const int digit = c - '0';
          if (digit > 9) return false;
          val = val * 10 + digit;
        }
        if (val > 255) return false;
      }
      return count == 4;
    };
    auto ipv6 = [&] () {
      if (queryIP.find(':') == string::npos) return false;
      while (getline(ip, cur, ':')) {
        ++count;
        if (count > 8 || cur.empty() || cur.size() > 4) return false;
        for (const char c : cur)
          if (!((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')))
            return false;
      }
      return count == 8;
    };

    if (ipv4()) return "IPv4";
    if (ipv6()) return "IPv6";
    return "Neither";
  }
};

class Solution_OLD {
 public:
  string validIPAddress(string IP) {
    if (IP.empty() || IP.back()=='.' || IP.back()==':')
      return "Neither";
    istringstream ip (IP);
    int count = 0;
    string res = "Neither", tmp = "";
    if (IP.find(':') == string::npos) {
      while (getline(ip, tmp, '.')) {
        ++count;
        if (4<count || tmp.empty() || (1<tmp.size() && tmp[0]=='0') || 3<tmp.size())
          return res;
        int val = 0;
        for (char c: tmp) {
          int tmpp = c-'0';
          if (9<tmpp)
            return res;
          val = val*10 + tmpp;
        }
        if (val<0 || 255<val)
          return res;
      }
      res = count == 4? "IPv4": "Neither";
    } else {
      while (getline(ip, tmp, ':')) {
        ++count;
        if (8<count || tmp.empty() || 4<tmp.size())
          return res;
        for (char c: tmp)
          if (!(('0'<=c && c<='9') || ('a'<=c && c<='f') || ('A'<=c && c<='F')))
            return res;
      }
      res = count == 8? "IPv6": "Neither";
    }
    return res;
  }
};

int main() {
  Solution s;
  
  return 0;
}

File: python3/468validateIpAddress.py
class Solution:
    def validIPAddress(self, queryIP: str) -> str:
        def isIPv4(s):
            try: return str(int(s)) == s and 0 <= int(s) <= 255
            except: return False
        def isIPv6(s):
            if len(s) > 4: return False
            try: return int(s, 16) >= 0
            except: return False

        if queryIP.count(".") == 3 and all(isIPv4(i) for i in queryIP.split(".")):
            return "IPv4"
        if queryIP.count(":") == 7 and all(isIPv6(i) for i in queryIP.split(":")):
            return "IPv6"
        return "Neither"

class Solution1:
    dic = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0', 'a', 'b', 'c', 'd', 'e', 'f',
           'A', 'B', 'C', 'D', 'E', 'F']
    def validIPAddress(self, queryIP: str) -> str:
        if '.' in queryIP:
            strs = queryIP.split(".")
            if len(strs) != 4:
                return "Neither"
            for str_num in strs:
                if not str_num.isdigit() or int(str_num) > 255 or \
                (len(str_num) > 1 and str_num[0] == '0'):
                    return "Neither"
            return "IPv4"
        elif ':' in queryIP:
            strs = queryIP.split(":")
            if len(strs) != 8:
                return "Neither"
            for part in strs:
                if len(part) == 0 or len(part) > 4:
                    return "Neither"
                for digit in part:
                    if digit not in self.dic:
                        return "Neither"
            return "IPv6"
        else:
            return "Neither"
